data/src/main/java/com/microsoft/azure/kusto/data/auth/ApplicationCertificateTokenProvider.java

```diff
@@ -5,18 +5,29 @@
 
 import com.microsoft.aad.msal4j.ConfidentialClientApplication;
 import com.microsoft.aad.msal4j.IClientCertificate;
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+
 import org.jetbrains.annotations.NotNull;
 
 import java.net.MalformedURLException;
 import java.net.URISyntaxException;
 
 public class ApplicationCertificateTokenProvider extends ConfidentialAppTokenProviderBase {
+    private final String applicationClientId;
+    private final IClientCertificate clientCertificate;
+
     ApplicationCertificateTokenProvider(@NotNull String applicationClientId, @NotNull IClientCertificate clientCertificate, @NotNull String clusterUrl, String authorityId) throws URISyntaxException {
         super(clusterUrl, authorityId);
+        this.applicationClientId = applicationClientId;

```

Comment:
Asaf, I'm not sure how this code ties up to the current PR, or if you are fixing auth issues on the way (where are those used?).

Reply:
It used to initalize the `clientApplication` object in the ctor, but now we can't (because we don't want a web request in a ctor), so the logic moved to `onCloudInit`.

That meant that I need to save the params that are later passed to the clientApplication ctor.